### ISSUE TITLE
Add support for resolving unregistered types

### ISSUE THREAD

**Issue by craigfowler** [csf-dev/CSF.Extensions.DependencyInjection#2]
This is a FlexDi feature I'd like to cover.  From https://github.com/csf-dev/FlexDi/issues/20.

**Comment by craigfowler** [comment]
This is more complex than I had imagined, because of the reasons described in cd88993134910d2ee5db1af13200f2330d480683.  I'm not confident that I can do this the way I'd like to.

However, I'm also wondering if I actually need this functionality either.  With libraries like **[Scrutor]**, it's very easy to bulk-register types for resolution as scoped services.  I'm not sure that I actually need to worry about resolving unregistered types, given that it's so easy to ensure everything is actually registered.

[Scrutor]: https://github.com/khellang/Scrutor

**Comment by craigfowler** [comment]
I've tried another approach today, in which I tried accepting the limitation of using `internal` methods.  However, there's another limitation which blocks me.  That's that the `IServiceProvider` which gets registered in DI [is always the base one](https://github.com/dotnet/runtime/blob/d7948dc7b9b061c7e3ee9ef6e3c9dd6854d195c0/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L63).  That means that when using the service provider, my decorator for it would never have been resolved.  So the whole approach is actually flawed anyway.

Decorator cannot be used to extend a service provider in this way.
